### ISSUE TITLE
Canvas texture expiry timing is unpredictable

### ISSUE THREAD

**Issue by kainino0x** [gpuweb/gpuweb#3295]
EDIT: Updated title to reflect both: issues with invisible canvases, and general unreliability of timing in the event loop.

---

Right now it's not clear whether [Update the rendering of the WebGPU canvas](https://gpuweb.github.io/gpuweb/#abstract-opdef-update-the-rendering-of-the-webgpu-canvas) gets called for canvases not in the DOM. This affects the advancement of the frame, i.e. whether the next call to `getCurrentTexture()` will return the same texture or a new one.

WebGL (with `preserveDrawingBuffer: false`) has an equivalent question, which is whether the drawing buffer gets replaced or not. This test case tries to test that: https://codepen.io/kainino0x/pen/mdxKKyM
Unfortunately, the results are kind of wonky.
- Firefox and Chrome seem to advance the frame only if the canvas is actually visible (inside the viewport) OR it's had transferControlToOffscreen called on it. Neither of them advances a `new OffscreenCanvas`.
- Safari seems to advances the frame as long as it's in the DOM. (no OffscreenCanvas support)

---

Tasks (see below)

- [ ] Figure out behavior for canvases
- [ ] Consider if behavior for external textures should match closer to behavior for canvases

**Comment by kainino0x** [comment]
I enhanced the test case: https://codepen.io/kainino0x/pen/mdxKKyM

I found one divergence in behavior between Chrome and Firefox, but they're still remarkably similar and I don't really understand the logic that leads to that.

Chrome on my Mac

![image](https://user-images.githubusercontent.com/606355/184034891-e9b4e873-17c4-4edf-9bc8-14348d47ed8a.png)

Firefox has one difference: instead shows a blank image in the second case, frame 3.

**Comment by kainino0x** [comment]
Actually, the transferControlToOffscreen cases are all flaky in both Chrome and Firefox.

**Comment by kenrussell** [comment]
On an M1 Mac, both Chrome Canary and Firefox Nightly seem to reliably produce the same results. Canary is the first screenshot here; Firefox is the second.

<img width="1107" alt="Frame Advancement M1 Chrome Canary" src="https://user-images.githubusercontent.com/1149904/184037483-7925c980-cb38-4230-998a-adad99e2b840.png">
<img width="1107" alt="Frame Advancement M1 Firefox Nightly" src="https://user-images.githubusercontent.com/1149904/184037503-93db6669-f3ee-44b4-9643-36afd585f259.png">

**Comment by kenrussell** [comment]
It would likely be impossible to tightly specify this behavior. The compositor isn't guaranteed to pick up and display every rendered frame, especially if the machine is slow or overloaded. https://registry.khronos.org/webgl/specs/latest/1.0/#2.2 doesn't tightly tie in to the defined HTML rendering algorithms, but it does tighly specify what operations cause the canvas to be conceptually "dirtied" and which will eventually lead to a frame advancement. Perhaps the WebGPU spec should focus on this aspect.

**Comment by kainino0x** [comment]
Ah, I see now the key thing: in WebGL it's undefined behavior. This also means there's no established WebGL behavior that we would want to try to match.

> If this flag is false, attempting to perform operations using this context as a source image after the rendering function has returned can lead to undefined behavior.

WebGPU has no `preserveDrawingBuffer: true` equivalent, so I don't think we can have undefined behavior since users would have no way to escape it.

**My proposal would be this, then:** all canvases and non-standalone `OffscreenCanvas`es<sup>1</sup> would get updated in the event loop of the thread they exist on, just after `rAF` (on the main thread, corresponding to when the `document` has its rendering update). `new OffscreenCanvas`es would never get automatically advanced, only in `transferToImageBitmap()`. In the common case you would probably instead just clear the resource with a render pass `"clear"`.

<sup>1</sup> I'm not precisely sure how to spec this for OffscreenCanvases transferred to workers, but I think it can be done.

**Comment by kainino0x** [comment]
@toji based on your recent work on this in Chromium, do you think this proposal makes sense?

**Comment by toji** [comment]
I'm a little worried about the difference in `OffscreenCanvas()` behavior, though I recognize it probably needs to be different to a certain degree.

Chrome's behavior right now, as of https://chromium-review.googlesource.com/c/chromium/src/+/3811084, is actually to "advance to the next frame" when the current task is completed. We talk about it in terms of "frames", but if you were to make a render loop with `setInterval(drawFrame, 1)` it would continue to function as expected (though there's no guarantee that every new texture would actually be composited.)

And honestly, I think that's probably a decent model? And crucially it applies equally well to both onscreen and offscreen canvases. `transferToImageBitmap()` needs a bit of special casing, but that's not too bad. Stated in simplified terms the logic would be: "The texture returned by `getCurrentTexture()` is valid until either the end of the current task or the drawing buffer is explicitly transferred."

Given that rAFs are all processed as part of a single task this gives most developers the functionality they expect, while not being overly dependent on the compositor and all the visibility questions it brings with it.

**Comment by toji** [comment]
Kai and I had a fairly lengthy chat conversation about the exact timing of when frames are advanced both in terms of the HTML spec and Chrome's implementation, which differs a bit from what the spec currently says.

After a lot of spec spelunking, reasoning about order of operations, and wondering what constitutes a "task" in Blink, we want to propose that we detach the current texture from it's backing drawing buffer any time a [task](https://html.spec.whatwg.org/multipage/webappapis.html#concept-task) in which `getCurrentTexture()` was called has been completed. This would include rAF callbacks, event callbacks, timeouts, etc.

For example, if you called `getCurrentTexture()` in a rAF callback, the texture would be valid for the remainder of the rAF callbacks as well as any microtasks scheduled during the rAF (like a `Promise.resolve()`), but then would immediately be detached so that the same texture couldn't be used in, say, a subsequent `setTimeout()`. (As stated in the spec currently, the drawing buffer WOULD remain available until the next `getCurrentTexture()` for read-only operations like copies to other contexts.)

The specific point in the spec we believe will achieve this best is in the [perform a microtask checkpoint](https://html.spec.whatwg.org/multipage/webappapis.html#perform-a-microtask-checkpoint) step. Specifically, we can mimic the [Cleanup Indexed Database Transactions](https://w3c.github.io/IndexedDB/#cleanup-indexed-database-transactions) step that is used by Indexed DB to execute logic after all of a tasks microtasks have been completed (effectively at the end of the task). That spec has a note that pretty well reflects what we are hoping for as well.

> These steps are invoked by [[HTML]](https://w3c.github.io/IndexedDB/#biblio-html). They ensure that [transactions](https://w3c.github.io/IndexedDB/#transaction-concept) created by a script call to [transaction()](https://w3c.github.io/IndexedDB/#dom-idbdatabase-transaction) are deactivated once the task that invoked the script has completed.

**Comment by kainino0x** [comment]
We have a tangential proposal to go with this one, too (alluded to in #3324), which is to move the "expire stale external textures" call to the same place. They're very similar APIs in the sense that they're both textures that get automatically `destroy()`ed by the UA.

Right now, stale import expiry happens in the event loop, which is very similar, but means it only gets checked for tasks that run at the toplevel of the event loop, not for event handlers, rAF callbacks, etc. If we end up putting canvas texture expiry in the microtask checkpoint, we could easily move stale import expiry to the same point.

This is NOT as important as the canvas issue because there's no fundamental issue with the way import expiry works now. The only semantic difference would be that stale imports would last slightly less long.

**Comment by kainino0x** [comment]
Back on the topic of canvas textures:

I'll go into some more detail about why we ended up proposing canvas textures expire after every task, instead of something closer to what we had previously described (expire after rAF callbacks when canvases get presented).

Naively, it seems to make sense to simply "advance the frame" (expire the current canvas texture) when the canvas "gets presented". However, this actually gets complex when you consider: there can be multiple documents (e.g. popup windows), and there can be multiple threads using the canvas.

There are a whole bunch of different configurations of canvases:

1. canvas element {in, not in} the DOM, controlled locally
2. canvas element {in, not in} the DOM, controlled by transferControlToOffscreen() from the main thread
3. canvas element {in, not in} the DOM, controlled by transferControlToOffscreen() from a worker
4. `new OffscreenCanvas` (not associated with any canvas element)

Starting with 4, `new OffscreenCanvas`. This one should probably be considered to be totally independent of the DOM and any presentation logic, so we'll just say this never implicitly expires. It only expires on `transferToImageBitmap()`.

Next, there are canvases which are controlled from the same thread (1 and 2) as the document they're on. No problem, they should be driven by the `requestAnimationFrame` of the window of the document they were created from. (e.g. `window.requestAnimationFrame` vs `popup.requestAnimationFrame`).

Finally, there are canvases that are being controlled from a worker. This is where things get iffy. There's only one `globalThis.requestAnimationFrame` in a worker, as far as I understand, and it's [defined](https://html.spec.whatwg.org/multipage/webappapis.html#event-loop-processing-model) as follows:

> If (...) the user agent believes that it would benefit from having its rendering updated at this time, then:
> - (...)
> - Run the animation frame callbacks for that DedicatedWorkerGlobalScope, passing in now as the timestamp.

This is technically fine for driving the page to render. The browser can just fire requestAnimationFrame every time it wants *any* of the canvases (that that worker has access to) to rerender.

But when it comes to expiry of canvas textures, we no longer have a way to rigorously define when expiry occurs. We can't just do it "every" frame, because there isn't a concept of a frame, and updates are driven entirely by this non-rigorous definition for rAF calls in the HTML spec. Hence, we came to the proposal that Brandon made, which we can rigorously define.

---

Another alternative to Brandon's proposal is to just unconditionally expire all canvas textures on the thread at some point in the event loop, rather than in the microtask checkpoint. I think both are equally sound.

**Comment by kenrussell** [comment]
This sounds like a reliable and well-testable semantic.

Would you anticipate any changes to the browser's ability to drop frames if it can't keep up with the application? I'm thinking about situations today where the compositor might skip running if it's falling behind its frame deadlines.

It's likely there's no significant change - the browser can still throttle `requestAnimationFrame`, and probably overall JavaScript execution in case applications are using discouraged patterns like rendering upon `setTimeout`. The compositor can decide to not display the frames it acquires from the WebGPU-rendered canvas. Also, the WebGPU implementation will internally throttle `getCurrentTexture` to prevent (as best as possible) applications from swamping the GPU.

**Comment by kainino0x** [comment]
This shouldn't change anything about the browser's ability to drop frames. Expiry of a canvas texture doesn't cause presentation; presentation always happens in ["Update the rendering"](https://html.spec.whatwg.org/multipage/webappapis.html#update-the-rendering), the frequency of which is completely controlled by the browser.

What this changes is that you can get many different canvas textures between two "Update the rendering" calls, when before you could only get one. However this should be fine, because all those extra ones just go into the void (get cleared and reused).

And yes, `getCurrentTexture` can always block if the application is running ahead too far by disobeying `requestAnimationFrame`.

**Comment by kainino0x** [comment]
We'll try to solicit some feedback about the options before the meeting. For summary they are, relative to the [event loop steps](https://html.spec.whatwg.org/multipage/webappapis.html#event-loop-processing-model):

1. Current thing: Expire only after rAF callbacks, between steps 7.13 and 7.16 (and fix spec to also expire after worker rAF between steps 9.1.2 and 9.1.3)
2. Middle option: Expire after top-level tasks (between steps 3 and 4) *and* after rAF callbacks
3. Option Brandon described: Expire in the [microtask checkpoint](https://html.spec.whatwg.org/multipage/webappapis.html#perform-a-microtask-checkpoint). This happens after each top-level task, each rAF callback, each rVFC callback, I think each event callback, etc.

**Comment by kainino0x** [comment]
Brandon and I spent a bit more time trying to think of the ergonomic impact of this change and had trouble identifying any. We think it's very hard to write an application that relies on any particular current canvas texture living across more than one task.

There are only a few cases where you really have control over this, e.g. you call requestAnimationFrame twice in the same frame, both callbacks should always be called in the same frame (thus would get the same texture in 1/2, different textures in 3).

There are various ways to enqueue tasks, e.g. `setTimeout(, 0)`, when called from rAF or otherwise from the same point in the rendering loop, where the tasks will *usually* will end up in the same frame, but it's not guaranteed. In this case, it's much safer to have reliable behavior (2/3) than behavior that will almost always work but then break when your setTimeout takes a long time to get scheduled (1).

**Comment by kainino0x** [comment]
> We'll try to solicit some feedback about the options before the meeting.

Feedback we got was that we should just enqueue a microtask from inside `getCurrentTexture()`. This has deterministic behavior because microtasks always execute in the order they're enqueued, so wouldn't cause any flakiness. You'd still see some microtasks happening before the expiration and others happening after, but in a predictable way. It's also within the usual execution model of the web (`queueMicrotask` is in the platform) and so is easier to understand at a glance for web developers experienced with tasks/microtasks.

So that is now our proposal.

**Comment by kainino0x** [comment]
As for video frames, we don't want the possibility of video frames expiring between rVFC callbacks and rAF callbacks, so probably don't want to change the behavior we have now. But I'll ask for feedback on that too.

**Comment by litherum** [comment]
I don't want to state any opinion (yet) about whether or not frames (`getCurrentTexture()`) should march onward for canvases that are not attached to the DOM.

However, if they _do_ march onward, microtask boundary is the wrong granularity. There is no browser-side rendering that ever happens between microtasks, so it doesn't make sense that we'd ever swap between microtasks. And it's totally legit for code to do things like call `.then()` on a promise that's already resolved and expect that any updates in the `then` block will be presented atomically with the rest of the page - this is already true of every DOM manipulation (and WebGL I believe).

**Comment by litherum** [comment]
Also, why would we ever swap more frequently than HTML5's Update the Rendering step? That seems wasteful.

**Comment by kainino0x** [comment]
Agreed let's defer on the topic of unobserved canvases for the moment. I think that is a reasonably separable discussion.

---

"Swap" is the wrong terminology here. We're not swapping a backbuffer and a frontbuffer. We're only terminating access to the drawing buffer - the compositor then picks up that buffer in "update the rendering", not immediately. Our proposed change _does_ allow applications to get multiple different canvas textures in the same frame, but that can be seen simply as overwriting the drawing buffer again.

The point here is to terminate write access at a more predictable time than what we have now. What we have now is fraught with unpredictable timing that will inevitably cause portability issues between different browsers, faster/slower devices, etc.

> And it's totally legit for code to do things like call `.then()` on a promise that's already resolved and expect that any updates in the `then` block will be presented atomically with the rest of the page - this is already true of every DOM manipulation (and WebGL I believe).

I technically agree that this is legit, but I have a very hard time thinking of a case where a page could ever rely on this being the case. When would the caller of a `.then()` or `await` reliably know that the Promise is already resolved and not pending?

And again, the point here is to make the behavior extremely well defined. Sure, the behavior might occasionally surprise a developer, but they'll never have a bug caused by timing that's difficult to reproduce or browser-dependent.

---

> It also strikes me that, overall, this is a very similar question to "do Custom Paint callbacks get run when the element is unobserved in the same ways?"

PaintWorklets don't run on the event loop - they provide programs (that happen to be defined in JS) that occur during browser paint. AFAICT there are no mechanisms for communicating between PaintWorklet and any other JavaScript or any other part of the outside world (except e.g. through timing side channels). So I don't think they help us answer this question.

Aside: It occurred to me that if there were a point where DOM layout changes became visible back to the application, that state change could be a good place for our state change as well. However I don't think there is such a place; observing layout just causes forced synchronous layout: https://www.afasterweb.com/2015/10/05/how-to-thrash-your-layout/

**Comment by litherum** [comment]
> When would the caller of a .then() or await reliably know that the Promise is already resolved and not pending?

As I mentioned in the meeting, there are 2 (popular) mechanisms today for doing this:

1. [`queueMicrotask()`](https://developer.mozilla.org/en-US/docs/Web/API/HTML_DOM_API/Microtask_guide) is a direct way to queue a micro task within the current task
2. `new Promise((resolutionFunction) => { ... })`. Calling the `resolutionFunction` synchronously schedules a microtask within the current task

I don't think we should discount the increasing prevalence of promises on the web platform, both as a way to model asynchronous completion, but also as a generally useful programming paradigm. I do think it is unreasonable to tell web authors they must have straight-line code (which was an argument brought up in today's call).

Being able to defer work like this is a very powerful, and very useful feature - it's the basis for the RAII pattern in C++. Swift has a similar `defer` feature too. Using promises in theses ways is both fairly common (in my experience) and likely to become increasingly common.

I also don't think it's impossible to come up with an invalidation schedule that is both A) more coarser-grained than microtasks, and B) is more well-defined than WebGL's. We just need to do the necessary research to determine what solutions are in that space.

**Comment by kainino0x** [comment]
Can you point to a real world sample, or at least provide a concrete example, of microtasks being used in this way? It's certainly possible, but I haven't seen it, and need to understand the usage better to be convinced that it makes sense to use in rendering code. To me it seems that trying to use microtasks like RAII or Swift's `defer` would be a messy misuse of async JavaScript for synchronous programming.

To me, microtasks seem fundamentally unsuitable for this "defer" pattern, which is a synchronous-programming primitive. Deferred work will get mixed with actual async work, nondeterministically (because promise callbacks can run either in the current iteration of the event loop, or a future one, depending on the relative timing of the promise resolution). Moreover, microtasks execute in forward order while `defer`/RAII execute in reverse order for good reasons. I think `defer` _is_ ideally what I'd want for canvas textures, but your argument has been precisely that microtasks *don't* work that way.

Now, that is a fair point - and if anything I think we should pull the expiry _sooner_, like in the first option below, so it works more like a `defer` would. To me, queuing a microtask is just a more well-lit approach that sufficiently approximates it.

Options I see here are:

- Expire the texture exactly at the end of the current task (like an "always first" microtask) (somehow)
- Queue a microtask to expire the texture
- Queue a microtask to queue a microtask to expire the texture (I don't think anybody likes this)
- Expire the texture at the end of the [microtask checkpoint](https://html.spec.whatwg.org/multipage/webappapis.html#perform-a-microtask-checkpoint) (like an "always last" microtask)
- Expire the texture at several specific points in [the event loop](https://html.spec.whatwg.org/multipage/webappapis.html#event-loop-processing-model). Either:
  - After step 3, and after step 7.13, and after step 9.1.2
  - After step 9, or equivalently before step 2

Here's a bonus idea that I don't expect anyone to like:

- getCurrentTexture cancels any existing cancellation microtask and makes a new one. This way you can call getCurrentTexture once at the beginning of a task so you can use it, then again a second time at the end of a task just to extend the lifetime past all microtasks enqueued in the task. Anytime during its lifetime you could repeatedly extend the lifetime past the current end of the microtask queue.

**Comment by greggman** [comment]
I'm going to throw out a 3rd option, unaware of what came up before. What if you had to explicitly release the texture? 

```
const tex = context.getCurrentTexture();

...

context.releaseCurrentTexture(tex);
```

Then you can hold on to it as long as you want.  If you call `getCurrentTexture` again and haven't released the previous one you get an error, if you try to release a different texture that one one you got from `getCurrentTexture` then you get an error.

**Comment by greggman** [comment]
Let me add, I don't know what the use cases are but, emscripten has asyncify and apparently wasm is adding a lower overhead way to implement it. I can certainly imagine as more apps are ported from native to webgpu via wasm they might end up having some asyncified features that end up causing issues here if their code does something that to them appears synchronous but emscripten/wasm has made it transparently async.

**Comment by kainino0x** [comment]
This would be an "explicit present" which we probably last discussed several years ago at one of the F2F meetings (I think San Diego 2019?). We had decided our swap should work similarly to existing canvases, though at that time didn't fully work through the fact that WebGL is the only thing that defines swap semantics at all (`preserveDrawingBuffer: false`) and that its timing was not fully defined.

I think there's a lot of value in an explicit present API, though I'm not sure now is the time to switch to one. We could add one later on (as canvas configuration option), though if people like the idea we could switch to it.

It does occur to me, now, that since we *can* make an improved model in the future, it wouldn't be so bad to keep the current behavior (expire only after rAF callbacks). After all, people are using the same model in WebGL today, and it's not the biggest issue. I'd love to do better, but it's not critical.

**Comment by toji** [comment]
> Here's a bonus idea that I don't expect anyone to like:
> 
> getCurrentTexture cancels any existing cancellation microtask and makes a new one.

Actually, this feels like a reasonable pattern? The scenario that it would fix is that you've got some dependency that calls `getCurrentTexture()` before your code gets into it's normal draw routine, where you will also call `getCurrentTexture()`. That's perfectly valid, but because another piece of code (possibly unknown to you) has [licked the current texture cookie first](https://devblogs.microsoft.com/oldnewthing/20091201-00/?p=15843) they now "own" its expiration, which makes it impossible for you to schedule a microtask during which the current texture is valid.

If, on the other hand, we allow a new call to `getCurrentTexture()` to extend the expiration by scheduling a new microtask you as the developer always have control over whether or not the current texture will remain valid for microtasks which you schedule, and it doesn't have any adverse effects on other callers. (You can never force the texture to outlive the task boundary, and the only scenario I can conceive of under which you would want to have forced the texture to expire before a particular microtask is in a unittest of the expiration behavior, which is easily controlled.)

I'm not sure if there's existing facilities to actually "cancel" a scheduled microtask, but that's not a big deal. We can just have the microtasks scheduled by `getCurrentTexture()` decrement a reference count and expire the texture when it hits zero.

**Comment by kainino0x** [comment]
Refcounting is a less weird way of describing it. Makes it sound better!

Brandon and I discussed, and support the refcounting-microtask option for V1. We'd like to later design an explicit present API post-V1, or at least something else that works well with WASM promise-integration (that is, engine-supported Asyncify).

**Comment by greggman** [comment]
I think I'm confused

```
  requestAnimationFrame(() => {
    getCurrentTexture()            // µtasklist: [expire]         expireCount = 1

    // do stuff with texture

    queueMicroTask(function foo() {
      getCurrentTexture();
      // do stuff with texture
    });                            // µtasklist: [expire, foo]    expireCount = 1

  });
```

now first task on µtasklist executes which is `expire` and expireCount is 1 which
gets decremented to 0, texture is expired

next `foo` gets executed, it calls `getCurrentTexture()` but the texture I wanted is already
expired.

**Comment by toji** [comment]
Yeah, this is probably the most confusing bit. The pattern developers would actually need to use is:

```
requestAnimationFrame(() => {
    getCurrentTexture()     // µtasklist: [expire]         expireCount = 1

    // do stuff with texture

    queueMicrotask(function foo() {
      // do stuff with texture
    });                    // µtasklist: [expire, foo]    expireCount = 1
    
    // Extend the expiration of the texture.
    getCurrentTexture();   // µtasklist: [expire, foo, expire]    expireCount = 2
});
```

Counterintuitive unless you are thinking about everything in terms of microtasks, and then it's at the very least predictable. I suspect any such usage of the API will be fairly advanced, so my concern would be making sure that the expiration is predictable and controllable, and not so much on making that pattern it perfectly intuitive. Some good error messaging in the case of using an expired texture would help significantly, I think.

**Comment by greggman** [comment]
> Some good error messaging in the case of using an expired texture would help significantly, I think.

I'm probably forgetting part of the spec but what does this mean? I'd expect `getCurrentTexture` called after expiring the previous current texture to get the next current texture, not an error.

**Comment by toji** [comment]
You're correct, `getCurrentTexture` would get the next texture at that point. But if you were to, say, take the previous texture handle and attach it as a render pass target, we'd have an opportunity to raise an error saying that access to the texture had expired.

**Comment by toji** [comment]
The editors talked more about this on Monday and while it was a worthwhile discussion we didn't reach any agreements. While Kai and I would have been fine with going with the microtask-base approach outlined above Myles was concerned about setting up a expiration model that diverges too far from the behavior we see in Canvas 2D, and argument that Kai and I do find compelling.

(You can skip this next paragraph if you only care about the proposal)

To clarify, Canvas 2D doesn't have an expiration in the same sense that WebGPU (and to a lesser degree WebGL) does, but it does have some pretty clear behavior about how the contents that you draw will show up on screen. Specifically, if you draw to a 2D canvas, then schedule a microtask in which you draw some more, you're implicitly guaranteed that you'll never get only the first half of that drawing, because rendering opportunities in the HTML spec only come after all microtasks have been completed. This would differ in WebGPU if we went with the microtask approach, since you could conceivably render half you scene, then schedule a microtaks where you render the other half of your scene, depending on the exact implementation details and order of event's it's feasible that you only get the latter half of your scene rendered. This CAN be worked around by careful developers, but it's a potential break in expectations nonetheless, _especially_ if we are sticking with an implicit present model for the express purpose of not upending developer expectations around how canvas works.

So, after some more discussion about alternative approaches, Kai and I have **yet another proposal**:

Canvas texture expiration can be scheduled as a task (not a microtask) on it's own task queue. That task queue can then be giving high priority by the implementation, so that tasks within it are dispatched prior to other tasks that may execute JavaScript, such as events or timers, ensuring that you only have write access to the texture for the space of the current task.

This relies on the fact that user agents are given flexibility to process tasks from various tasks queue in the order and priority that they see fit. Specifically in the [Processing Model](https://html.spec.whatwg.org/multipage/webappapis.html#event-loop-processing-model) algorithm it states:

> 1. Let taskQueue be one such [task queue](https://html.spec.whatwg.org/multipage/webappapis.html#task-queue), chosen in an [implementation-defined](https://infra.spec.whatwg.org/#implementation-defined) manner.

An example block within the [task queue](https://html.spec.whatwg.org/multipage/webappapis.html#task-queue) definition expands on the reasoning for this:

> For example, a user agent could have one [task queue](https://html.spec.whatwg.org/multipage/webappapis.html#task-queue) for mouse and key events (to which the [user interaction task source](https://html.spec.whatwg.org/multipage/webappapis.html#user-interaction-task-source) is associated), and another to which all other [task sources](https://html.spec.whatwg.org/multipage/webappapis.html#task-source) are associated. Then, using the freedom granted in the initial step of the [event loop processing model](https://html.spec.whatwg.org/multipage/webappapis.html#event-loop-processing-model), it could give keyboard and mouse events preference over other tasks three-quarters of the time, keeping the interface responsive but not starving other task queues. Note that in this setup, the processing model still enforces that the user agent would never process events from any one [task source](https://html.spec.whatwg.org/multipage/webappapis.html#task-source) out of order.

Given that this is explicitly implementation defined behavior, it wouldn't be appropriate for WebGPU to strictly require that the expiration tasks always take priority, but we could use language along the lines of the following to nudge implementations towards more predictable scheduling:

> Implementations **should** always execute tasks in the [=WebGPU expiry task queue=] before task queues for events, timers, etc. Since WebGPU expiry tasks can only be enqueued from JavaScript code, there is no possibility of scheduling starvation.

This would give us:

  - Canvas textures that remain valid _at least_ through the lifetime of the current task
  - Use of existing spec mechanisms without injecting new WebGPU-specific steps into the event model
  - Less strict predictability than a microtask-based approach but more intuitive predictability for the majority of web devs

Thoughts?

**Comment by litherum** [comment]
Sounds great!

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/gpu-web-2022-10-05#canvas-texture-expiry-timing-is-unpredictable-3295">GPU Web 2022-10-05</a></summary>

* CW: Brandon you made a proposal.
* BJ: first, title in agenda not quite right. Canvas texture expiry in general.
    * (KN: fixed)
* BJ: previously: wanted these to expire on microtask boundary. Kai, Myles and I discussed. Myles' argument persuasive. If implicit presentation - should match expectations coming from other canvas APIs. Tricky because those aren't perfect fit for WebGPU. Hard to do lot of Canvas 2D rendering with part in microtask, and have only half of it show up. Implicit guarantee - from event processing, etc. - your drawing in tasks+microtasks will show up together. Unfortunate if we break this for WebGPU.
* Summary of [comment in the github issue](https://github.com/gpuweb/gpuweb/issues/3295#issuecomment-1268751595):
    * BJ: iterated on this. This AM - seems reasonable - came up with this:
    * BJ: if you look at task processing in HTML spec, everything gets put on its own task queue. Things grouped together like keyboard input. Reason: browsers should be able to give queues priority as necessary. So that mouse/keyboard input can have higher priority for example.
    * BJ: what if expiration was scheduled as a task on its own texture expiration queue? Strongly encourage browsers to give this fairly high priority. (Is Kelsey laughing? :) )
    * BJ: probably inappropriate that this task queue gets higher priority than others. Reasonable to say - we encourage browsers to do this with reasonably high priority. Specifically, encouraged to execute them before the next thing that'll create JS callbacks.
    * BJ: then - texture lives for entirety of task. Nothing about weird refcounting with microtasks. It'll be guaranteed to live for your task. Expired soon after.
    * BJ: pretty lightweight for browsers to handle.
    * BJ: doesn't require us to invent new language/mechanisms. Using well understood mechanisms, and nudging browsers toward a priority preference.
* KG: all to avoid commit()?
* BJ: yes, to some degree. Preference to say - we'd like to introduce explicit swap later. Discussed in past - don't want to do anything breaking current patterns too drastically for canvas APIs. Wanted implicit commit first. This is a way to get there in a better specified way. Post V1, can discuss explicit commit. Think it'd be valuable. Esp as we see more ports via WebAssembly.
* KN: also as we see Wasm Promise Integration proposal come online.
* KG: concerned that it's different from what WebGL does. 
* BJ: that's a fundamental problem. WebGL's not well spec'd in this area. Technically is probably different from what WebGL impls do - but for side effects for users, gets closer than microtask-oriented solutions.
* KN: think it's not too far off from WebGL. Forgot - we'd queue a task that expires the texture, and also when we present it. Expectation, those two will look the same to users. They can't tell the difference between something happening in a task or between tasks. In WebGL, spec says roughly - when you exit the task, not guaranteed you can still see the texture. That's because of the way the event loop works. Between every task there's a possibility of presentation. By also expiring in a task, we make it expire regardless of whether we present it. And it does so sooner. If you get a texture early in the frame, it'll get terminated early. Can't rely on that in WebGL per spec, so it's about the same thing.
* KG: but we're making a new thing that probably expires things earlier. Want roughly to do the same thing that WebGL does. Another option - if you want implicit presentation, we do something different? Trying to 
* KR: HTML spec authors gave us feedback not to do commit() on OffscreenCanvas. We’re almost there on this spec. Let’s do something that is better defined without a big change. I don’t think it’s hard to spec or implement.
* KG: making a new task queue…don't know how to do that.
* BJ: most specs do that. We say "WebXR input events task queue" for example. That's it. Becomes one of the group of task queues that are processed.
* KG: my POV - you're saying do harder than an API that works already. More accurate, but may break more content. Impl's something I'm not familiar with and can't comment on.
* BJ: see where you're coming from. My perspective a little different - we have behavior most people are OK with, but is ill-deifned. Can we use language of web specs to …
* …more discussion, too much interruption…
* BJ: think with task priorities we get close to what you're talking about.
* KG: think you're spec'ing something poorly defined that's not testable.
* BJ: think you're saying, spec it poorly defined because that's the way we did it in the past. Trying to find approach we can take which is less reliant on formal spec language, and achieve the same thing.
* KR: breaking things on task boundary will let you test - for example setTimeout(0) can assert that you get a new texture.
* BJ: want something reasonable, predictable, works across the board. Expiring at task boundaries gives us it in whatever web context you want.
* KG: want to hear Apple's perspective.
* CW: can you talk with other Mozillians?
* KG: that's not the issue - think this s less useful.
* BJ: can you give a test case this would break?
* KG: … think it’s more likely to break things
* KN: think it’s equally likely.
* CW: WebGL, at random times can get a new texture. Idea is - in WebGPU, try to define it - you can't render for more than one task. Sorry. Instead of apps being flakily broken when the compositor runs - make WebGPU consistent, no, you can't do this because it breaks sometimes.
* KG: think it'll be no less flaky. Want Myles' feedback.
* KN: think that's accurate. Browsers should do this. Not guaranteed to be less flaky. But an expectation that we'd try.
* RC: if someone uses a lot of microtasks and draws in each one, does the browser compositor have an obligation to show each one? Or just the last?
* BJ: good to clarify. In this new proposal, we're talking about tasks, not microtasks. If you have a task where you getCurrentTexture(), and schedule a bunch of microtasks - and call getCurrentTexture in each, and draw something new - under this proposal, you're guaranteed to get back the same texture for every microtask part of the same task. If you do load instead of clear for color attachment - all those microtasks can accumulate content onto the same surface. Once task's completed, opportunity to do the rendering. Something Myles was very concerned about being guaranteed. Didn't want difference in behavior - expiring on microtask meant you could get texture in microtask did some drawing, and now you get a new texture in a subsequent microtask. *Not* a danger in this proposal - addressing Myles' concern.
* KN: never an obligation to present oen of these textures. Can always skip it and present the next one.
* RC: so this is talking about which texture you get back, the same one or a cleared one. And if you get a cleared one, browser's still not obligated to show all of them.
* KN: exactly. Bunch of tasks in one frame - each one will get a new texture, and browser might only show the last one.
* BJ: similar to WebGL. Can render in mousemove - you can draw bunch of content that might not be drawn.
* RC: so for this proposal, for mousemove - will you get back a new texture each time?
* BJ: proposal will say, the browser should. Hard to require. May enforce restrictions on something in the core HTML spec. Reasonable interpretation - you'd get new texture in each task.
* KN: think most of what we're talking about is similar to Myles' suggestion to expire after microtasks are completed. Similar things. Just different ways of expressing it.
* RC: low-latency canvas / desynchronized: if you put this in, does it decouple things even more? Do what you want with this texture?
* KN: don't think we'll do desynchronized the same way as in WebGL. similar but not exactly.
* CW: need Apple to get consensus anyway - need to get back to this. Would like to get to next agenda item.

</details>

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/gpu-web-meeting-2022-10-12-13-apac-timed#canvas-texture-expiry-timing-is-unpredictable-3295">GPU Web Meeting 2022-10-12/13 APAC-timed</a></summary>

* KN: previously: KG had concerns about clamping things down more significantly than WebGL, instead of relying on WebGL experience - where things seem to sort of be working OK.
* KG: discussed within Moz - felt that this the tricky part of the problem, would be nicely solved by explicit present option. People wouldn't have to worry about it.
* MM: in far future - looking forward to first-class support for variable refresh rate displays on the web. Right now, requestVideoFrame, and you watch a 24 FPS video - you get the 3:2 judder. Reason - web was designed with fixed framerate in mind. rVF could be driven by something more fundamental than screen refresh rate.
* MM: little worried that if web platform adds VRR support - if WebGPU adds explicit present, these would fight each other. Synchronized animations across multiple APIs is important.
* JB: how does explicit present ensure there's a conflict between WebGPU & VRR presentation of content?
* MM: I'm expecting that explicit present would take arguments. "Present this WebGPU frame at this time in the future with this duration.". Then this functionality comes to rAF, and they fight.
* KG: don't see why you couldn't implement VRR in rAF framework today.
* KN: API I envision for explicit present has none of that. Just saying you're done writing the frame. Don't think it conflicts with refresh rate considerations.
* KN: I think explicit present is a great possible direction. Also not the ideal API for everyone. Think there are a lot of use cases benefiting from it, like Wasm Promise integration. Frame spanning multiple rAFs. Want to build something more basic right now. Don't think having explicit present is the right API for everyone. Have to be able to do present at the end of your frame. Not easy to fit into all existing (web) engines. Would like to keep present API more similar to WebGL/2D canvas. And want that API, since it's the default, to have consistent behavior so it's easy to debug for developers. That's why I'd like to take the approach we've proposed - find a good deterministic point to terminate write access to the texture. Will look like WebGL/2D, and have repeatable behavior.
* KG: worried about how it changes vs. how WebGL does. Interested in pushing this to WebGL too?
* KR: Web compat concern with WebGL. Have encountered applications that relied on undefined behavior.
* KN: think we could do something like this for WebGL. As long as it conforms closely to what browsers are doing.
* KG: **would like to know more about the flickering problems you mentioned with WebGL. Hard for us to make an informed decision**.
    * KR: for others: we should look up the bug from Scirra with an async rAF callback - discussed that one already - there were others, let's discuss offline. They're all in Chromium's issue tracker somewhere.

</details>

**Comment by greggman** [comment]
something that's come up recently trying to work on debugging tools that might inform this discussion....

Ideally I'd like debugging tools to show "frames". The definition of a "frame" appears to be all the commands you issued until you called `getCurrentTexture` plus whatever commands come after that until the texture expires. It's got no relation to requestAnimationFrame nor setTimeout or anything else.

Given the current spec, how would someone who wants to capture all the commands wrap WebGPU in such a way as to capture these frames?

There are hacks, you can base the capture on rAF but then of course it doesn't work if the page is generating frames with setTimeout, or requestVideoFrame, or some other myriad of possible events

I was thinking the debug wrapper could "queue a microtask to end capture for the current frame" when `getCurrentTexture` is called with the microtask ending the frame. That's not quite right though based on the current direction the spec is headed. 

Is this something that should inform the design or when a frame ends or is this just kind of "too bad, if you want the actual correct ending of a frame for capture you'll need some special private API", or maybe I'm just not thinking of the right solution. Maybe a `setTimeout(endCaptureForCurrentFrame, 0)` since a timeout is a task, not a microtask?

**Comment by kainino0x** [comment]
I don't have a perfect answer to this. I'm unsure whether it needs to inform the design.

How I'd do this depends on exactly what you want to capture. If you want to cut boundaries when the canvas is "snapshotted" for the screen, then rAF should be appropriate, regardless of what method is used to drive rendering. Expiry has to be guaranteed to happen by the time the canvas is presented (right after rAF callbacks), otherwise there's a race condition. Using this method there could be more than 1 frame of the application between cuts, if rendering is driven faster than rAF, but frames shouldn't get cut in half if the canvas is for display. It's trickier if the canvas is being used for something else, like if it's going into a video encoder, or being used by the application as an intermediate image, but rAF may still work.

I think theoretically you could also detect cuts based on what actually returns from getCurrentTexture. If it's the same frame, the object will be `===`, if not, it'll be a new object. Of course, the getCurrentTexture call itself can easily be "late" in the frame, so it's not the right cut point. But depending on what kind of debugging you're doing, for example if you're recording a command stream, you could use it as a signal to retroactively insert a cut point somewhere between every pair of getCurrentTexture calls that return different values. If you have some way to detect where the beginning of the current task was in the stream, perhaps, you can go back and insert it there.

You could also detect expiry by repeatedly checking whether the texture has been destroyed. Doing this is kind of ugly, but possible: in a tight task loop of some kind, pushErrorScope, try to `createView()` from the texture last returned last returned from getCurrentTexture, popErrorScope, repeat. The result is asynchronous so you'd still need to insert the cut somewhere earlier in the stream. It's also at slight risk of heisenbugs since it actually issues WebGPU calls.

**Comment by greggman** [comment]
I don't think you can call `getCurrentTexture` and check for `===` as you've now likely broken the user's code. `getCurrentTexture` causes something to happen (the texture will now expire and/or be presented even if you weren't planning on using it).

I guess you could wrap it and see if when the user has called it, has it changed since last time. On the other hand, that's not really a signal that they ended the frame. 

```
cmd
cmd
cmd
getCurentTexture
cmd
cmd
cmd
getCurentTexture
cmd
cmd
cmd
```

Which commands belong to one frame and which to another?

`getCurrentTexture` happens in the middle of preparing a frame (you might computing shadows, deferred rendering,  etc...), before calling `getCurrentTexture) but all of those are relevant part of "capturing the current frame".

It worse if the user is rendering to multiple canvaes with the same device.

**Comment by kainino0x** [comment]
Yeah, I should have clarified, I was thinking of hooking getCurrentTexture, not calling it at all. But yes, it doesn't give you the correct signal by itself. So it's not that good.

Editors discussed this in a bit more depth, and we are pretty sure that the current proposal still provides something on par with what you can do today with WebGL. That is, if the application is driven by rAF, you can capture based on rAF, and if it isn't, then capturing based on rAF isn't perfect but should still work. The concept of a "frame" is not that well defined for non-rAF applications anyway.

**Comment by litherum** [comment]
Note: There are HTML PRs that add steps to the "update the rendering" steps that are domain-specific. E.g. here's a PR that adds a new step in the "update the rendering" steps to fixup element focus: https://github.com/whatwg/html/pull/8392

So this approach of adding a new step is something we could do, too.

**Comment by kainino0x** [comment]
I forgot to write stuff specifying how it works for offscreen canvases, which I believe is:

- if it's not transferred (created by `new OffscreenCanvas`), expiry only happens on `transferToImageBitmap`. Readbacks happen against the current texture, are implicitly synchronized, and do not expire the current texture. (With a copy needed if the alphaMode is `opaque` which is unfortunately the default but largely useless for `new OffscreenCanvas`es.)
- if it's a `transferControlToOffscreen` `OffscreenCanvas`, expiry happens in (no later than) the "Update the rendering of that dedicated worker to reflect the current state" step of the worker event loop.
